hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mob/mapreduce/MobRefReporter.java

```diff
@@ -0,0 +1,485 @@
+ *         area for this table and CF, but it is kept there because a _different_ table has a
+ *         reference to it (e.g. from a snapshot clone). If these other tables are removed then
+ *         the file will likely be deleted unless there is a snapshot also referencing it.
+ *   * ARCHIVE BUT NO HLINKS - the reference is currently present in the archive for this table and
+ *         CF, but there are no references present to prevent its removal. Unless it is newer than
+ *         the general TTL (default 5 minutes) or referenced in a snapshot it will be subject to
+ *         cleaning.
+ *   * ARCHIVE BUT IOE WHILE CHECKING HLINKS - Check the job logs to see why things failed while
+ *         looking for why this file is being kept around.
+ *   * MISSING FILE - We couldn't find the reference on the FileSystem. Note that MOB cells contain
+ *         a pointer to the table the reference was originally created in, much like HFileLinks do,
+ *         but that pointer is implemented via a server-side tag. It is thus possible that lookups
+ *         to impacted rows will still work (e.g. if the HFileLink in our table's mob area was
+ *         manually removed, but the original file in the referenced table is still around for some
+ *         reason.) However, even in this edge case HBase's internal cleaning systems should
+ *         eventually remove that file once it is no longer needed by another table or snapshot.
+ *   * HLINK BUT POINT TO MISSING FILE - There is a pointer in our mob area for this table and CF
+ *         to a file elsewhere on the FileSystem, however the file it points to no longer exists.
+ *   * MISSING FILE BUT IOE WHILE CHECKING HLINKS - We could not find the referenced file, however
+ *         you should check the job logs to see why we couldn't check to see if there is a pointer
+ *         to the referenced file in our archive or another table's archive or mob area.
+ *
+ */
+@InterfaceAudience.Private
+public class MobRefReporter extends Configured implements Tool {
+  private static Logger LOG = LoggerFactory.getLogger(MobRefReporter.class);
+  public static final String NAME = "mobrefs";
+  static final String REPORT_JOB_ID = "mob.report.job.id";
+  static final String REPORT_START_DATETIME = "mob.report.job.start";
+  static final int SCAN_CACHING = 10000;
+  static final long ONE_DAY = 24 * 60 * 60 * 1000;
+
+  public static class MobRefMapper extends TableMapper<Text, ImmutableBytesWritable> {
+    @Override
+    public void map(ImmutableBytesWritable r, Result columns, Context context) throws IOException,
+        InterruptedException {
+      if (columns == null) {
+        return;
+      }
+      Cell[] cells = columns.rawCells();
+      if (cells == null || cells.length == 0) {
+        return;
+      }
+      Set<String> files = new HashSet<>();
+      long count = 0;
+      long size = 0;
+      for (Cell c : cells) {
+        if (MobUtils.hasValidMobRefCellValue(c)) {
```

Comment:
I personally don't think so, but I don't feel strongly here. We should output a log message if it's not a valid ref cell since we're supposed to be using this with a scan that asks for only ref cells.

Reply:
Ah nice if there's something to counter the `if` that helps my eye.